Bonsai.PulsePal/PulsePal.cs

```diff
@@ -56,28 +57,79 @@ public bool IsOpen
             get { return serialPort.IsOpen; }
         }
 
-        void serialPort_DataReceived(object sender, SerialDataReceivedEventArgs e)
+        Task RunAsync(CancellationToken cancellationToken)
         {
-            var bytesToRead = serialPort.BytesToRead;
-            if (serialPort.IsOpen && bytesToRead > 0)
+            serialPort.Open();
+            serialPort.ReadExisting();
+            Connect();
+
+            return Task.Factory.StartNew(() =>
             {
-                bytesToRead = serialPort.Read(readBuffer, 0, bytesToRead);
-                for (int i = 0; i < bytesToRead; i++)
+                using var cancellation = cancellationToken.Register(serialPort.Dispose);
+                while (!cancellationToken.IsCancellationRequested)
                 {
-                    ProcessInput(readBuffer[i]);
+                    try
+                    {
+                        var bytesToRead = serialPort.BytesToRead;
+                        if (bytesToRead == 0)
+                        {
+                            var nextByte = serialPort.ReadByte();
+                            if (nextByte < 0) break;
```

Comment:
Do we want to break here if the end of the stream is read? Is it not possible to have more bytes to read after this?

Reply:
@RoboDoig reading the end of stream means the serial port is closed. From the docs the call to [ReadByte](https://learn.microsoft.com/en-us/dotnet/api/system.io.ports.serialport.readbyte?view=dotnet-plat-ext-8.0) is blocking, so it either waits until a new byte is read, or if it returns with -1 it means `Dispose` was called.

This is actually an important  component for this pattern to work correctly cross-platform, it turns out that `ReadByte` is one of the few interruptible synchronous blocking call with predictable behavior on Mono. More details can be found in https://github.com/bonsai-rx/bonsai/pull/1444 when we first used this pattern for both Arduino and Harp serial ports.